File: Data Structures & Algorithms/combination-target-sum/submission-1.java
class Solution {

    public void dfs(int i, int currSum, List<List<Integer>> result, List<Integer> combo, int[] nums, int target)
    {
        if(i == nums.length && currSum == target)
        {
            result.add(new ArrayList<Integer>(combo));
            return;
        }
        if(currSum > target || i == nums.length)
        {
            return;
        }

        currSum = nums[i] + currSum;
        combo.add(nums[i]);
        dfs(i, currSum, result, combo, nums, target);
        currSum -= nums[i];
        combo.remove(combo.size()-1);
        dfs(i + 1, currSum, result, combo, nums, target);
    }

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> combo = new ArrayList<>();
        dfs(0, 0, result, combo, nums, target);
        return result;
    }
}

File: Data Structures & Algorithms/combination-sum-iv/submission-1.py
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = [0] * (target + 1)
        dp[0] = 1
        def dfs(total):
            if dp[total] != 0:
                return dp[total]
            for n in nums:
                if n > total:
                    continue
                dp[total] += dfs(total - n)
            return dp[total]
        return dfs(target)
            

